arrow-core-test/src/main/kotlin/arrow/test/laws/FoldableLaws.kt

```diff
@@ -1,42 +1,114 @@
+      Law("Foldable Laws: reduceLeftOption consistent with reduceLeftToOption") { FF.`reduceLeftOption returns Option value`(GEN, EQOptionInt) },
+      Law("Foldable Laws: reduceRightOption consistent with reduceRightToOption") { FF.`reduceRightOption returns Option value`(GEN, EQOptionInt) },
+      Law("Foldable Laws: traverse_ consistent with foldRight") { FF.`traverse consistent with foldRight`(Id.applicative(), GEN, Id.genK(), Id.eqK().liftEq(Eq.any())) },
+      Law("Foldable Laws: sequence_ consistent with traverse_") { FF.`sequence consistent with traverse`(Id.applicative(), GENK, Id.genK(), Id.eqK().liftEq(Eq.any())) },
+      Law("Foldable Laws: isEmpty returns if there are elements or not") { FF.`isEmpty returns if there are elements or not`(GEN, EQBool) },
+      Law("Foldable Laws: isNotEmpty consistent with isEmpty") { FF.`isNotEmpty consistent with isEmpty`(GEN, EQBool) },
+      Law("Foldable Laws: foldMapM folds on F mapping values to G(B) using given Monoid") { FF.`foldMapM folds on F mapping values to G(B) using given Monoid`(GEN, EQForIdInt) },
+      Law("Foldable Laws: get gets the item at the given index of the Foldable") { FF.`get gets the item at the given index of the Foldable`(GEN, EQOptionInt) }
+    )
+  }
+
+  fun <F> laws(FF: Foldable<F>, GA: Applicative<F>, EQK: EqK<F>): List<Law> {
+    val EQKListKInt: Eq<Kind<F, Kind<ForListK, Int>>> = EQK.liftEq(ListK.eq(Int.eq()))
 
     return listOf(
-        Law("Foldable Laws: Left fold consistent with foldMap") { FF.leftFoldConsistentWithFoldMap(GEN, EQ) },
-        Law("Foldable Laws: Right fold consistent with foldMap") { FF.rightFoldConsistentWithFoldMap(GEN, EQ) },
-        Law("Foldable Laws: Exists is consistent with find") { FF.existsConsistentWithFind(GEN) },
-        Law("Foldable Laws: Exists is lazy") { FF.existsIsLazy(GEN, EQ) },
-        Law("Foldable Laws: ForAll is lazy") { FF.forAllIsLazy(GEN, EQ) },
-        Law("Foldable Laws: ForAll consistent with exists") { FF.forallConsistentWithExists(GEN) },
-        Law("Foldable Laws: ForAll returns true if isEmpty") { FF.forallReturnsTrueIfEmpty(GEN) },
-        Law("Foldable Laws: FirstOption returns None if isEmpty") { FF.firstOptionReturnsNoneIfEmpty(GEN) },
-        Law("Foldable Laws: FirstOption returns None if predicate fails") { FF.firstOptionReturnsNoneIfPredicateFails(GEN) },
-        Law("Foldable Laws: FoldM for Id is equivalent to fold left") { FF.foldMIdIsFoldL(GEN, EQ) }
-      )
+      Law("Foldable Laws: orEmpty consistent with just empty") { FF.`orEmpty consistent with just empty`(GA, EQKListKInt) }
+    )
   }
 
+  fun <F> Foldable<F>.`foldRight is lazy`(G: Gen<Kind<F, Int>>, EQ: Eq<Int>) =
+    forAll(G) { fa: Kind<F, Int> ->
+      val sideEffect = SideEffect()
+      fa.foldRight(Eval.now(0)) { _, _ ->
+        sideEffect.increment()
+        Eval.now(1)
+      }.value()
+      val expected = if (fa.isEmpty()) 0 else 1
+      sideEffect.counter.equalUnderTheLaw(expected, EQ)
```

Comment:
I am not quite sure what exactly this is supposed to test. It executes the `Eval` returned by `foldRight` directly so it is not lazy at all. You should assign the eval to a value, check if the counter is still at 0 and then call eval and verify it being at 1.

Reply:
Also `equalUnderTheLaw` is a bit too much when comparing integers, just use the standart matchers like `should` directly.